File: mcp/cursor-state-bridge/state_io.py
"""Bridge-side state IO that imports the packaged workflow-state API.

The bridge writes workspace runtime JSON files, but it executes only the
package-owned implementation shipped under ``src/oh_my_cursor/workflow_state``.
It never importlib-loads Python from a workspace ``.cursor/state/`` directory.

This module performs zero ``subprocess`` calls -- the bridge never shells out
to the CLI. All write tools route through the library API.
"""
from __future__ import annotations

import importlib
import json
import sys
from pathlib import Path
from typing import Any

from jail import resolve_jailed


_WORKFLOW_STATE_MODULE_KEY = "_omcs_workflow_state"


_PAYLOAD_ROOT = Path(__file__).resolve().parents[2]
_SRC_DIR = _PAYLOAD_ROOT / "src"


# ---------------------------------------------------------------------------
# Library loader (imports package-owned workflow-state API once per process)
# ---------------------------------------------------------------------------


def _load_workflow_state(_workspace: Path):
    """Return the packaged workflow-state API module.

    ``_workspace`` is accepted for the existing handler contract; the
    executable implementation is resolved from the shipped payload, not from
    the runtime workspace.
    """
    if _WORKFLOW_STATE_MODULE_KEY in sys.modules:
        return sys.modules[_WORKFLOW_STATE_MODULE_KEY]

    api_path = _SRC_DIR / "oh_my_cursor" / "workflow_state" / "api.py"
    if not api_path.is_file():
        raise RuntimeError(f"workflow-state package not found at {api_path}")
    if str(_SRC_DIR) not in sys.path:
        sys.path.insert(0, str(_SRC_DIR))

    module = importlib.import_module("oh_my_cursor.workflow_state.api")
    module_file = getattr(module, "__file__", None)
    if module_file is None or Path(module_file).resolve() != api_path.resolve():
        raise RuntimeError(
            "workflow-state API resolved outside the trusted payload: "
            f"{module_file!r}; expected {api_path}"
        )
    sys.modules[_WORKFLOW_STATE_MODULE_KEY] = module
    return module


# ---------------------------------------------------------------------------
# Path resolution helpers
# ---------------------------------------------------------------------------


def _resolve_state_path(workspace: Path, task_id: str | None) -> Path:
    """Return the workflow-state.json path for a ``task_id`` (or default)."""
    if task_id:
        target = workspace / "docs" / "plans" / task_id / "workflow-state.json"
    else:
        target = workspace / ".cursor" / "state" / "workflow-state.json"
    # Validate jail containment before any IO and use the resolved path so
    # writer sidecars are created beside the real jailed state file.
    return resolve_jailed(workspace, target)


def _mcp_text(payload: Any) -> dict[str, Any]:
    """Wrap a Python value into the MCP ``content[].text`` response shape."""
    if isinstance(payload, str):
        text = payload
    else:
        text = json.dumps(payload, ensure_ascii=False)
    return {"content": [{"type": "text", "text": text}]}


# ---------------------------------------------------------------------------
# Tool implementations -- each one resolves the file path inside the jail,
# acquires the shared file_lock via the library API, and returns the
# written state wrapped in the MCP response shape.
# ---------------------------------------------------------------------------


def state_read(workspace: Path, params: dict[str, Any]) -> dict[str, Any]:
    """Read workflow-state.json (or return ``no state`` when absent)."""
    workspace_str = params.get("workspace")
    active_workspace = Path(workspace_str).resolve() if workspace_str else workspace
    target = _resolve_state_path(active_workspace, params.get("task_id"))
    if not target.exists():
        return _mcp_text("no state")
    try:
        parsed = json.loads(target.read_text(encoding="utf-8"))
    except json.JSONDecodeError as exc:
        raise RuntimeError(f"state file parse error: {exc}") from exc
    return _mcp_text(parsed)


_DEFAULT_HISTORY_CAP = 1000


def _history_cap(params: dict[str, Any]) -> int:
    """Extract optional ``history_cap`` from tool params (Phase 7).

    Defaults to 1000 when absent.  Non-int / invalid values fall back to
    the default rather than raise -- the cap is a retention knob, not a
    schema field.
    """
    raw = params.get("history_cap")
    if isinstance(raw, int):
        return raw
    if isinstance(raw, str) and raw.lstrip("-").isdigit():
        try:
            return int(raw)
        except ValueError:
            return _DEFAULT_HISTORY_CAP
    return _DEFAULT_HISTORY_CAP


def state_init(workspace: Path, params: dict[str, Any]) -> dict[str, Any]:
    task_id = params.get("task_id")
    if not isinstance(task_id, str) or not task_id:
        raise ValueError("state_init: task_id is required")
    target = _resolve_state_path(workspace, task_id if params.get("scope_per_task", True) else None)
    library = _load_workflow_state(workspace)
    state = library.init_state(
        target,
        task_id=task_id,
        title=str(params.get("title", "")),
        phase=str(params.get("phase", "intake")),
        status=str(params.get("status", "pending")),
        role=str(params.get("role", "orchestrator")),
        next_action=str(params.get("next_action", "")),
        history_cap=_history_cap(params),
    )
    return _mcp_text(state)


def state_set_phase(workspace: Path, params: dict[str, Any]) -> dict[str, Any]:
    task_id = params.get("task_id")
    if not isinstance(task_id, str) or not task_id:
        raise ValueError("state_set_phase: task_id is required")
    phase = params.get("phase")
    if not isinstance(phase, str) or not phase:
        raise ValueError("state_set_phase: phase is required")
    target = _resolve_state_path(workspace, task_id)
    library = _load_workflow_state(workspace)
    state = library.set_state(
        target,
        phase=phase,
        status=params.get("status") if isinstance(params.get("status"), str) else None,
        role=params.get("role") if isinstance(params.get("role"), str) else None,
        next_action=params.get("next_action") if isinstance(params.get("next_action"), str) else None,
        note=str(params.get("note", "advanced phase via cursor-state-bridge")),
        history_cap=_history_cap(params),
    )
    return _mcp_text(state)


def state_record_failure(workspace: Path, params: dict[str, Any]) -> dict[str, Any]:
    task_id = params.get("task_id")
    if not isinstance(task_id, str) or not task_id:
        raise ValueError("state_record_failure: task_id is required")
    message = params.get("message")
    if not isinstance(message, str):
        message = ""
    retry_count = params.get("retry_count", 0)
    if not isinstance(retry_count, int):
        try:
            retry_count = int(retry_count)
        except (TypeError, ValueError):
            retry_count = 0
    target = _resolve_state_path(workspace, task_id)
    library = _load_workflow_state(workspace)
    state = library.record_failure(
        target,
        type=str(params.get("type", "fixable")),
        message=message,
        retry_count=retry_count,
        note=str(params.get("note", "")),
        history_cap=_history_cap(params),
    )
    return _mcp_text(state)


def state_update_acceptance_criterion(workspace: Path, params: dict[str, Any]) -> dict[str, Any]:
    """Add or update an acceptance criterion.  Evidence stays OPTIONAL (R4)."""
    task_id = params.get("task_id")
    if not isinstance(task_id, str) or not task_id:
        raise ValueError("state_update_acceptance_criterion: task_id is required")
    ac_id = params.get("ac_id")
    if not isinstance(ac_id, str) or not ac_id:
        raise ValueError("state_update_acceptance_criterion: ac_id is required")
    status = params.get("status")
    if not isinstance(status, str) or not status:
        raise ValueError("state_update_acceptance_criterion: status is required")
    target = _resolve_state_path(workspace, task_id)
    library = _load_workflow_state(workspace)
    state = library.update_acceptance_criterion(
        target,
        ac_id=ac_id,
        status=status,
        criterion=params.get("criterion") if isinstance(params.get("criterion"), str) else None,
        evidence=params.get("evidence") if isinstance(params.get("evidence"), str) else None,
        note=str(params.get("note", "")),
        history_cap=_history_cap(params),
    )
    return _mcp_text(state)


def state_history_append(workspace: Path, params: dict[str, Any]) -> dict[str, Any]:
    """Append a free-form history entry without mutating top-level fields."""
    task_id = params.get("task_id")
    if not isinstance(task_id, str) or not task_id:
        raise ValueError("state_history_append: task_id is required")
    note = params.get("note") or params.get("event")
    if not isinstance(note, str) or not note:
        raise ValueError("state_history_append: note (or event) is required")
    target = _resolve_state_path(workspace, task_id)
    library = _load_workflow_state(workspace)
    state = library.append_history(
        target,
        note=note,
        phase=params.get("phase") if isinstance(params.get("phase"), str) else None,
        status=params.get("status") if isinstance(params.get("status"), str) else None,
        history_cap=_history_cap(params),
    )
    return _mcp_text(state)

File: mcp/cursor-state-bridge/state_io.ts
import * as fs from 'node:fs';
import * as path from 'node:path';
import { resolveJailed } from './jail.ts';
import { JailError } from './jail.ts';
import * as api from '../../src/oh_my_cursor/workflow_state/api.ts';

function failEscape(message: string): JailError {
  return new JailError(message);
}

function containsTraversalSegment(input: string): boolean {
  return input.split(/[\\/]+/).includes('..');
}

function activeWorkspace(workspace: string, params: any): string {
  const workspaceReal = fs.realpathSync(path.resolve(workspace));
  const workspaceOverride = params.workspace;
  if (workspaceOverride === undefined || workspaceOverride === null || workspaceOverride === '') {
    return workspaceReal;
  }
  if (typeof workspaceOverride !== 'string') {
    throw new Error('state_read: workspace must be a string when provided');
  }
  if (containsTraversalSegment(workspaceOverride)) {
    throw failEscape(`state workspace override contains traversal: ${workspaceOverride}`);
  }
  const requested = path.isAbsolute(workspaceOverride)
    ? path.resolve(workspaceOverride)
    : path.resolve(workspaceReal, workspaceOverride);
  let requestedReal: string;
  try {
    requestedReal = fs.realpathSync(requested);
  } catch {
    requestedReal = requested;
  }
  const relativeToWorkspace = path.relative(workspaceReal, requestedReal);
  if (relativeToWorkspace.startsWith('..') || path.isAbsolute(relativeToWorkspace)) {
    throw failEscape(`state workspace override escapes configured workspace: ${workspaceOverride}`);
  }
  return requestedReal;
}

function resolveStatePath(workspace: string, taskId: string | null | undefined): string {
  const ws = path.resolve(workspace);
  let target: string;
  if (taskId) {
    target = path.join(ws, 'docs', 'plans', taskId, 'workflow-state.json');
  } else {
    target = path.join(ws, '.cursor', 'state', 'workflow-state.json');
  }
  return resolveJailed(ws, target);
}

function mcpText(payload: any): any {
  const text = typeof payload === 'string' ? payload : JSON.stringify(payload);
  return { content: [{ type: 'text', text }] };
}

const DEFAULT_HISTORY_CAP = 1000;

function getHistoryCap(params: any): number {
  const raw = params.history_cap;
  if (typeof raw === 'number') {
    return raw;
  }
  if (typeof raw === 'string' && /^-?\d+$/.test(raw)) {
    try {
      return parseInt(raw, 10);
    } catch {
      return DEFAULT_HISTORY_CAP;
    }
  }
  return DEFAULT_HISTORY_CAP;
}

export function state_read(workspace: string, params: any): any {
  const target = resolveStatePath(activeWorkspace(workspace, params), params.task_id);
  if (!fs.existsSync(target)) {
    return mcpText('no state');
  }
  let parsed: any;
  try {
    const raw = fs.readFileSync(target, 'utf-8');
    parsed = JSON.parse(raw);
  } catch (err: any) {
    throw new Error(`state file parse error: ${err.message}`);
  }
  return mcpText(parsed);
}

export function state_init(workspace: string, params: any): any {
  const taskId = params.task_id;
  if (typeof taskId !== 'string' || !taskId) {
    throw new Error('state_init: task_id is required');
  }
  const scopePerTask = params.scope_per_task !== undefined ? params.scope_per_task : true;
  const target = resolveStatePath(workspace, scopePerTask ? taskId : null);
  const state = api.initState(target, {
    task_id: taskId,
    title: params.title !== undefined ? String(params.title) : undefined,
    phase: params.phase !== undefined ? String(params.phase) : undefined,
    status: params.status !== undefined ? String(params.status) : undefined,
    role: params.role !== undefined ? String(params.role) : undefined,
    next_action: params.next_action !== undefined ? String(params.next_action) : undefined,
    history_cap: getHistoryCap(params),
  });
  return mcpText(state);
}

export function state_set_phase(workspace: string, params: any): any {
  const taskId = params.task_id;
  if (typeof taskId !== 'string' || !taskId) {
    throw new Error('state_set_phase: task_id is required');
  }
  const phase = params.phase;
  if (typeof phase !== 'string' || !phase) {
    throw new Error('state_set_phase: phase is required');
  }
  const target = resolveStatePath(workspace, taskId);
  const state = api.setState(target, {
    phase: phase,
    status: typeof params.status === 'string' ? params.status : null,
    role: typeof params.role === 'string' ? params.role : null,
    next_action: typeof params.next_action === 'string' ? params.next_action : null,
    note: typeof params.note === 'string' ? params.note : 'advanced phase via cursor-state-bridge',
    history_cap: getHistoryCap(params),
  });
  return mcpText(state);
}

export function state_record_failure(workspace: string, params: any): any {
  const taskId = params.task_id;
  if (typeof taskId !== 'string' || !taskId) {
    throw new Error('state_record_failure: task_id is required');
  }
  let message = params.message;
  if (typeof message !== 'string') {
    message = '';
  }
  let retryCount = params.retry_count;
  if (typeof retryCount !== 'number') {
    if (typeof retryCount === 'string') {
      try {
        retryCount = parseInt(retryCount, 10);
      } catch {
        retryCount = 0;
      }
    } else {
      retryCount = 0;
    }
  }
  const target = resolveStatePath(workspace, taskId);
  const state = api.recordFailure(target, {
    type: typeof params.type === 'string' ? params.type : 'fixable',
    message,
    retry_count: retryCount,
    note: typeof params.note === 'string' ? params.note : undefined,
    history_cap: getHistoryCap(params),
  });
  return mcpText(state);
}

export function state_update_acceptance_criterion(workspace: string, params: any): any {
  const taskId = params.task_id;
  if (typeof taskId !== 'string' || !taskId) {
    throw new Error('state_update_acceptance_criterion: task_id is required');
  }
  const acId = params.ac_id;
  if (typeof acId !== 'string' || !acId) {
    throw new Error('state_update_acceptance_criterion: ac_id is required');
  }
  const status = params.status;
  if (typeof status !== 'string' || !status) {
    throw new Error('state_update_acceptance_criterion: status is required');
  }
  const target = resolveStatePath(workspace, taskId);
  const state = api.updateAcceptanceCriterion(target, {
    ac_id: acId,
    status: status,
    criterion: typeof params.criterion === 'string' ? params.criterion : null,
    evidence: typeof params.evidence === 'string' ? params.evidence : null,
    note: typeof params.note === 'string' ? params.note : undefined,
    history_cap: getHistoryCap(params),
  });
  return mcpText(state);
}

export function state_history_append(workspace: string, params: any): any {
  const taskId = params.task_id;
  if (typeof taskId !== 'string' || !taskId) {
    throw new Error('state_history_append: task_id is required');
  }
  const note = params.note || params.event;
  if (typeof note !== 'string' || !note) {
    throw new Error('state_history_append: note (or event) is required');
  }
  const target = resolveStatePath(workspace, taskId);
  const state = api.appendHistory(target, {
    note: note,
    phase: typeof params.phase === 'string' ? params.phase : null,
    status: typeof params.status === 'string' ? params.status : null,
    history_cap: getHistoryCap(params),
  });
  return mcpText(state);
}
